### DISCUSSION TITLE
Function Framed::next truncates message

### DISCUSSION THREAD

**Discussion by hanusek** [tokio-rs/tokio#5463] [Q&A]
**Version**
1.25.0

**Platform**
Ubuntu 22.04.1 LTS
kernel: 5.15.0-53-generic 

**Description**
I created it based on the example [chat](https://github.com/tokio-rs/tokio/blob/master/examples/chat.rs) - custom server.
I use [BytecCodec](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.BytesCodec.html) and [Framed](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html).

## short summary of the bug

I tried this code:
```rust
let acceptor = TlsAcceptor::from(cfg.clone());
let tls_stream = acceptor.accept(stream).await.unwrap();
let peer = Peer::new(state.clone(), Framed::new(tls_stream, BytesCodec::new())).await?;
```
https://github.com/hanusek/tokio-tls-example/blob/websocket2/server/src/main.rs#L190

Sometimes, the function [next()](https://docs.rs/futures-util/0.3.26/futures_util/stream/trait.StreamExt.html#method.next) returns truncated bytes.

**I expected to see this happen:** 
Read One message with 960.

**Instead, this happened:**
After the first reading, I have a message with 192 bytes. And after the following reading message with 768 bytes.

**Comment by Darksonn** [comment]
You seem to misunderstand how `tokio_util::codec` works. When using the `BytesCodec`, there is no such thing as a "message". It's just a sequence of individual bytes that it will put into chunks with the "chunk boundaries" chosen arbitrarily.

Perhaps you meant to use a different codec?

**Comment by hanusek** [comment]
> You seem to misunderstand how `tokio_util::codec` works. When using the `BytesCodec`, there is no such thing as a "message". It's just a sequence of individual bytes that it will put into chunks with the "chunk boundaries" chosen arbitrarily.
> 
> Perhaps you meant to use a different codec?

Thanks for your reply. 
I would like to send and receive structs (serialized by the [speedy](https://crates.io/crates/speedy) library).
In the first version not used struct [Framed](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.Framed.html) and all was ok. 
How to fix it?

**Reply by Darksonn** [reply]
Well, your code calls the variable `lines`, so if the messages are separated by newline characters, you can use the [`LinesCodec`](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.LinesCodec.html).

**Reply by hanusek** [reply]
> Well, your code calls the variable `lines`, so if the messages are separated by newline characters, you can use the [`LinesCodec`](https://docs.rs/tokio-util/latest/tokio_util/codec/struct.LinesCodec.html).

No. The name - 'lines' is incorrect. I based on the example with LinesCodec but my protocol is binary. Now, I use ByteCodec.

**Reply by Darksonn** [reply]
The codec must be written to work with whatever protocol you are using. You can see an example for how to implement your own codec [in the module documentation](https://docs.rs/tokio-util/latest/tokio_util/codec/index.html).

**Reply by hanusek** [reply]
Maybe I should implement a custom Framed? I guess the function read works incorrectly. 
I found a similar question: https://users.rust-lang.org/t/in-tokio-trying-to-handle-multiple-message-sources-in-one-task/48189
